portal/models/questionnaire_response.py

```diff
@@ -523,16 +535,48 @@ def __init__(self, user, qb_id):
             QuestionnaireResponse.encounter_id).order_by(
             QuestionnaireResponse.authored)
 
-        self.qnrs = []
+        self._qnrs = []
         for qnr in query:
-            self.qnrs.append(QNR(
+            self._qnrs.append(QNR(
+                qnr_id=qnr.id,
                 qb_id=qnr.questionnaire_bank_id,
                 iteration=qnr.qb_iteration,
                 status=qnr.status,
                 instrument=qnr.instrument_id.split('/')[-1],
                 authored=qnr.authored,
                 encounter_id=qnr.encounter_id))
 
+    def completed_qs(self, qb_id, iteration):
+        """Return set of completed Questionnaire results for Indefinite"""
+        # ignore the qb_id - as prior protocol versions picked up also count
+        return {
+            qnr.instrument for qnr in self.qnrs
+            if qnr.status == "completed"}
+
+    def partial_qs(self, qb_id, iteration):
+        """Return set of partial Questionnaire results for Indefinite"""
+        # ignore the qb_id - as prior protocol versions picked up also count
+        return {
```

Comment:
Should these return lists or filtered Queryset for consistency?

Reply:
no - sigh - it's a bit of a mess.  there are comments elsewhere explaining how required returns a list as it's ordered (only makes sense for non indef, but you get the idea) where the others return sets for easy set method checks.